contracts/src/finance/vesting_wallet.rs

```diff
@@ -0,0 +1,621 @@
+        self.start() + self.duration()
+    }
+
+    #[selector(name = "released")]
+    fn released_eth(&self) -> U256 {
+        self._released.get()
+    }
+
+    #[selector(name = "released")]
+    fn released_erc20(&self, token: Address) -> U256 {
+        self._erc20_released.get(token)
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_eth(&self) -> U256 {
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        self.vested_amount_eth(block::timestamp()) - self.released_eth()
+    }
+
+    #[selector(name = "releasable")]
+    fn releasable_erc20(
+        &mut self,
+        token: Address,
+    ) -> Result<U256, Self::Error> {
+        let vested = self.vested_amount_erc20(token, block::timestamp())?;
+        // SAFETY: total vested amount is by definition greater than or equal to
+        // the released amount.
+        Ok(vested - self.released_erc20(token))
+    }
+
+    #[selector(name = "release")]
+    fn release_eth(&mut self) -> Result<(), Self::Error> {
+        let amount = self.releasable_eth();
+        let released = self
+            .released_eth()
+            .checked_add(amount)
+            .expect("should not exceed `U256::MAX` for `_released`");
+        self._released.set(released);
+
+        evm::log(EtherReleased { amount });
+
+        transfer_eth(self.ownable.owner(), amount)
+            .map_err(|_| Error::ReleaseEtherFailed(ReleaseEtherFailed {}))
+    }
+
+    #[selector(name = "release")]
+    fn release_erc20(&mut self, token: Address) -> Result<(), Self::Error> {
+        let amount = self.releasable_erc20(token)?;
+        let owner = self.ownable.owner();
+
+        self.safe_erc20.safe_transfer(token, owner, amount).map_err(|_| {
+            Error::ReleaseTokenFailed(ReleaseTokenFailed { token })
+        })?;
+
+        // SAFETY: total supply of a [`crate::token::erc20::Erc20`] cannot
+        // exceed `U256::MAX`.
+        self._erc20_released.setter(token).add_assign_unchecked(amount);
+
+        evm::log(ERC20Released { token, amount });
+
+        Ok(())
+    }
+
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_eth(&self, timestamp: u64) -> U256 {
+        let total_allocation = contract::balance()
+            .checked_add(self.released_eth())
+            .expect("total allocation should not exceed `U256::MAX`");
+
+        self.vesting_schedule(total_allocation, U64::from(timestamp))
+    }
+
+    #[selector(name = "vestedAmount")]
+    fn vested_amount_erc20(
+        &mut self,
+        token: Address,
+        timestamp: u64,
+    ) -> Result<U256, Self::Error> {
+        if !Address::has_code(&token) {
+            return Err(InvalidToken { token }.into());
+        }
+
+        let erc20 = IErc20::new(token);
+        let call = Call::new_in(self);
+        let balance = erc20
+            .balance_of(call, contract::address())
+            .expect("should return the balance");
```

Comment:
I believe you do not need to check `Address::has_code(&token)` but just map an error into `InvalidToken` here. `Erc20::balance_of` should always return `U256`. Additionally, any other contract may exist on this address but it may not be `Erc20` - it will panic and this path is not covered in rust docs.